### ISSUE TITLE
Incorrect "missing match arm" error

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust-analyzer#4896]
The code [linked here](https://github.com/rust-lang/rust/blob/ff4a2533a0720f9cdd86e02eafa3725f07aa7752/src/librustc_mir/interpret/intern.rs#L106) produces an incorrect "missing match arm" error. The missing arm would be `MemoryKind::Machine(x)`, but it is not needed because `x: !`.

**Comment by jonas-schievink** [comment]
Note that this requires `#![feature(exhaustive_patterns)]` to be accepted by rustc

**Comment by CAD97** [comment]
This happens equivalently with an empty `enum Void {}`, if the type is not known. If the type is found, the problem is not emmited.

Minimal examples:

```rust
enum Void {}
fn test(v: Void) {
    match v {}
}
```

This works properly, as r-a sees the `Void` type.

```rust
enum Either<A, B> {
    A(A),
    B(B),
}
fn sorry_ra() {
    enum Void {}
    match Either::<(), Void>::A(()) {
        Either::A(()) => (),
        Either::B(void) => match void {},
    }
}
```

This shows a problem on the `void` of `match void`, claiming a type of `{unknown}` and a missing match arm.

```
enum Either<A, B> {
    A(A),
    B(B),
}
enum Void {}
fn sorry_ra() {
    enum Void {}
    match Either::<(), Void>::A(()) {
        Either::A(()) => (),
        Either::B(void) => match void {},
    }
}
```

This makes the shown problem go away, as `void` is now typed as `Void` (but the wrong one).

-----

I think this manifestation of this issue boils down to r-a not (yet?) supporting item definitions in statement positions. I think the correct resolution to both this version and incidentally the `!` version is to just not emit the missing match arm problem when the type being matched over is unknown.

**Comment by bjorn3** [comment]
> I think this manifestation of this issue boils down to r-a not (yet?) supporting item definitions in statement positions.

It doesn't support item definitions in inside functions: #1165

**Comment by CAD97** [comment]
The stable case with `{unknown}` uninferred types was addressed by #5787.

**Comment by RalfJung** [comment]
I am getting more "missing match arm" messages, which do not seem to involve `exhaustive_patterns` (at least I don't think this code relies on that flag):

https://github.com/rust-lang/rust/blob/d7c338641466d54bf8d4b2eae5d6865483e1d3f4/compiler/rustc_mir/src/interpret/intrinsics.rs#L73

https://github.com/rust-lang/rust/blob/d7c338641466d54bf8d4b2eae5d6865483e1d3f4/compiler/rustc_mir/src/interpret/validity.rs#L517

I could not find a way for RA to tell me *which* match arm it thinks is missing.

---

Is there a way to have RA not report missing match arms? I'd rather wait a bit for `cargo check` to be done in exhcange for actually being able to rely on all the errors I see being meaningful.

**Comment by Veykril** [comment]
Closed by #13167
